lib/retry_ratio_tracker.js

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+module.exports = RetryRatioTracker;
+
+var assert = require('assert');
+var RateCounter = require('./rate_counter');
+
+var RATE_INTERVAL = 1000;
+var DEFAULT_RATE_COUNTER_NUM_OF_BUCKETS = 20;
+
+function RetryRatioTracker(options) {
+    assert(options.timers, 'options.timers required');
+
+    if (!(this instanceof RetryRatioTracker)) {
+        return new RetryRatioTracker(options);
+    }
+
+    var numOfBuckets = options.numOfBucketsInRateCounter || DEFAULT_RATE_COUNTER_NUM_OF_BUCKETS;
+    var rateCounterOpts = {
+        numBuckets: numOfBuckets
+    };
+
+    this.timers = options.timers;
+    this.retryRateCounter = RateCounter(rateCounterOpts);
+    this.requestRateCounter = RateCounter(rateCounterOpts);
+    this.refreshCounterInterval = RATE_INTERVAL / numOfBuckets;
+
+    this.refreshRateCounters();
+}
+
+RetryRatioTracker.prototype.currentRetryRatio = function currentRetryRatio() {
+    if (this.requestRateCounter.rps === 0) {
```

Comment:
This logic will work create once the process is stable because it will look at the last 20 * 50ms buckets for both data.

However it may behave weird around startup time.

For example a service might do 100 QPS, but in the first 50ms we may do 5 requests and try to do 4 retries, with 5 requests we are allowed 0.5 retries but we actually have budget for 10 retries.

Given the implementation of `this.requestRateCounter` we can actually add another gaurd.

```
// Whilst the request rate counter is not yet warmed up, allow ALL retries
if (this.requestRateCounter.buckets.length < numOfBuckets) {
  return 0;
}
```

I don't know what the best approach is here to be honest in the warmup phase, maybe something more dump like 

```
// Whilst the request rate counter is not warmed up, allow a "free" 10 retries
if (this.requestRateCounter.buckets.length < numOfBuckets &&
    this.retryRateCounter.rps < 10
) {
  return 0;
}
```